src/Wallet.ts

```diff
@@ -0,0 +1,252 @@
+import * as bip39 from 'bip39';
+import * as bip32 from 'bip32';
+import * as bitcoinjs from 'bitcoinjs-lib';
+import coinSelectAccumulative, { Target } from 'coinselect/accumulative';
+import { BlockchainService, Utxo } from './BlockchainService';
+
+export type AddressMetadata = {
+  address: string;
+  derivationPath: string;
+  publicKey: Buffer;
+};
+
+export type WalletStats = {
+  balance: number;
+  pendingBalance: number;
+  nextUnusedAddress: AddressMetadata;
+  nextUnusedChangeAddress: AddressMetadata;
+  addresses: AddressMetadata[];
+  changeAddresses: AddressMetadata[];
+  addressesWithUtxo: AddressMetadata[];
+};
+
+export type UtxoWithMetadata = Utxo & {
+  address: string;
+  derivationPath: string;
+};
+
+export type AddressCollectionStats = Omit<
+  WalletStats,
+  'nextUnusedChangeAddress' | 'changeAddresses'
+>;
+
+export class Wallet {
+  constructor(
+    private blockchainService: BlockchainService,
+    private gapLimit = 20,
+    private derivationPath = "m/84'/0'/0'",
+  ) {}
```

Comment:
The Wallet is mostly stateless. In most modern web applications, state management is handled through dedicated libraries like Redux, so the user is expected to maintain the state of the Wallet separately. It's still easy to use, in my opinion, as you can see in the tests.